cvxpy/lin_ops/canon_backend.py

```diff
@@ -927,7 +1214,64 @@ def create_new_tensor_view(self, variable_ids: set[int], tensor: Any, is_paramet
         pass  # noqa
 
 
-class ScipyTensorView(TensorView):
+class DictTensorView(TensorView, ABC):
+    def accumulate_over_variables(self, func: Callable, is_param_free_function: bool) \
+            -> TensorView:
+        """
+        Apply 'func' to A and b.
+        If 'func' is a parameter free function, then we can apply it to all parameter slices
+        (including the slice that contains non-parameter constants).
+        If 'func' is not a parameter free function, we only need to consider the parameter slice
+        that contains the non-parameter constants, due to DPP rules.
+        """
+        for variable_id, tensor in self.tensor.items():
+            self.tensor[variable_id] = self.apply_to_parameters(func, tensor) if \
+                is_param_free_function else func(tensor[Constant.ID.value])
+
+        self.is_parameter_free = self.is_parameter_free and is_param_free_function
+        return self
+
+    @staticmethod
+    def combine_potentially_none(a: dict | None, b: dict | None) -> dict | None:
+        if a is None and b is None:
+            return None
+        elif a is not None and b is None:
+            return a
+        elif a is None and b is not None:
+            return b
+        else:
+            return TensorView.add_dicts(a, b)
+
+    @staticmethod
+    @abstractmethod
+    def add_tensors(a: Any | None, b: Any | None):
+        pass
+
+    @staticmethod
+    def add_dicts(self, a: dict, b: dict) -> dict:
+        """
+        Addition for dict-based tensors.
+        """
+        res = {}
+        keys_a = set(a.keys())
+        keys_b = set(b.keys())
+        intersect = keys_a & keys_b
+        for key in intersect:
+            if isinstance(a[key], dict) and isinstance(b[key], dict):
+                res[key] = NumpyTensorView.add_dicts(a[key], b[key])
+            elif isinstance(a[key], list | np.ndarray) and isinstance(b[key], list | np.ndarray):
```

Comment:
You can even add another abstractmethod that defines what type the tensor is.
So you add
```py
    @staticmethod
    @abstractmethod
    def tensor_type():
        pass
```
which is just `return list` for scipy and `return np.ndarray` for numpy.

Then you simplify to
```py
isinstance(a[key], self.tensor_type()) and isinstance(b[key], self.tensor_type()):
```

Reply:
ok neat, I was just thinking of way I could generalise the is instance to avoid having many "or".